server/app/survey_response.go

```diff
@@ -131,6 +161,49 @@ func (a *UserSurveyApp) sendAcknowledgementPost(userID, surveyPostID string) err
 	return nil
 }
 
+func (a *UserSurveyApp) matchSurveyAndResponse(survey *model.Survey, response *model.SurveyResponse) error {
+	// response can't have more answers than the number of questions in the survey
+	if len(response.Response) > len(survey.SurveyQuestions.Questions) {
+		return errors.New("incorrect number of responses submitted")
+	}
+
+	// if only one response is submitted, it needs to be
+	// the answer to the linear scale question
+	if len(response.Response) == 1 {
+		linearScaleQuestionID, err := survey.GetSystemRatingQuestionID()
+		if err != nil {
+			return err
+		}
+
+		if _, ok := response.Response[linearScaleQuestionID]; !ok {
+			return errors.New("linear scale question must be answered")
+		}
+
+		// When user selects a rating and submits via the Submit button,
+		// the client passes the response type manually, and we should only verify it,
+		// not override it.
+		if response.ResponseType != model.ResponseTypeComplete {
+			response.ResponseType = model.ResponseTypePartial
+		}
+	} else {
+		// make sure answered questions belong to the survey
+		surveyQuestionIDMap := map[string]bool{}
+		for _, question := range survey.SurveyQuestions.Questions {
+			surveyQuestionIDMap[question.ID] = true
+		}
+
+		for responseQuestionID := range response.Response {
+			if _, ok := surveyQuestionIDMap[responseQuestionID]; !ok {
+				return errors.New("invalid question ID found in submitted answer")
+			}
+		}
+
+		response.ResponseType = model.ResponseTypeComplete
+	}
```

Comment:
@harshilsharma63 this defers from the previous implementation in that the linear response no longer seems required in the response length is > 1. 

Isn't that the opposite of what's desired? (always check that the linear question is answered)

Reply:
This is what happens if you send a response with complete responseType and without the linear question

![Screenshot 2024-05-20 at 15 08 50](https://github.com/mattermost/mattermost-plugin-user-survey/assets/17114347/dc1f4c8c-be86-4587-be41-0e4d7b81ed34)

Reply:
The issue is that it's marked as a valid response, in the survey stats.